### ISSUE TITLE
InitializeTenancyByDomain $onFail logic not working

### ISSUE THREAD

**Issue by sauravkumal** [archtechx/tenancy#670]
#### Describe the bug
I tried InitializeTenancyByDomain $onFail logic as mentioned in document
```
InitializeTenancyByDomain::$onFail = function () {
    abort(404,'Oops ! This page has gone Missing');
};
```
but i still get the `Stancl\Tenancy\Exceptions\TenantCouldNotBeIdentifiedOnDomainException` error.
#### Steps to reproduce
followed steps from https://tenancyforlaravel.com/docs/v3/configuration/#static-properties

#### Expected behavior
as with the above code, the exception should be handled with a 404 abort.

#### Your setup
 - Laravel version: 8.1.2
 - stancl/tenancy version: 3.4

**Comment by stancl** [comment]
The logic works just fine https://github.com/archtechx/tenancy/blob/3.x/tests/DomainTest.php#L99

You probably used the code at a wrong location. Use a service provider `boot()` method.

**Comment by plakhin** [comment]
This doesn't work in case `Stancl\Tenancy\Features\UniversalRoutes::class` is enabled in config

**Comment by stancl** [comment]
Yeah for that you have to use a custom onFail handler which combines the two. Should be covered in one of the issues/discussions

I might make the logic more intelligent in v4

**Comment by plakhin** [comment]
Just a mention in docs would be enough, I think.
I'm not using universal routes, just found out the reason and answered here.

**Comment by aeq-dev** [comment]
Hello @stancl 
I'm facing the same issue, could you please help us in this : "use a custom onFail handler which combines the two "
thanks

**Comment by andreasmatuszewski** [comment]
> Yeah for that you have to use a custom onFail handler which combines the two. Should be covered in one of the issues/discussions
> 
> I might make the logic more intelligent in v4

Hi @stancl, 
i cannot find the issue you mentioned. I searched through all TenantCouldNotBeIdentifiedOnDomainException / InitializeTenancyByDomain $onFail logic issues.

Would appreciate your help on this. I am using universal routes only for livewire on central domain.

**Comment by sha-hin** [comment]
What I've done to solve the issue if someone still get into this because if enabled universal routes, I've to extend the UniversalRoutes Feature like that:
```
namespace App\Tenant\Features;

use \Stancl\Tenancy\Features\UniversalRoutes as BaseUniversalRoutes;
use Stancl\Tenancy\Tenancy;

class UniversalRoutes extends BaseUniversalRoutes
{
    public function bootstrap(Tenancy $tenancy): void
    {
        foreach (static::$identificationMiddlewares as $middleware) {
            $middleware::$onFail ??= function ($exception, $request, $next) {
                if (static::routeHasMiddleware($request->route(), static::$middlewareGroup)) {
                    return $next($request);
                }

                throw $exception;
            };
        }
    }
}
```

Then used the new UniversalRoutes feature on tenancy config file, finally the fail logic will be working as expected.

**Comment by stancl** [comment]
We changed this for v4 — https://github.com/archtechx/tenancy/commit/233a1222bf72769c28edcb7558fef7bf3ef216a0

**Comment by MCKLtech** [comment]
I get slightly weird behaviour when I use this code in V3:

```
<?php

namespace App\Http\Middleware\Tenancy;

use Stancl\Tenancy\Features\UniversalRoutes as BaseUniversalRoutes;
use Stancl\Tenancy\Tenancy;

class UniversalRoutes extends BaseUniversalRoutes
{

    public function bootstrap(Tenancy $tenancy): void
    {
        foreach (static::$identificationMiddlewares as $middleware) {
            $middleware::$onFail ??= function ($exception, $request, $next) {

                if (static::routeHasMiddleware($request->route(), static::$middlewareGroup)) {
                    return $next($request);
                }

                throw $exception;
            };
        }
    }
}
```

It seems to stop Livewire from working. When I use a Livewire component and view the network tab, it seems to return a 404 for updates. Livework works correctly with the standard UniversalRoutes.

This is my **TenancyServiceProvider**

```
public function boot()
    {
        $this->bootEvents();
       
        InitializeTenancyByDomain::$onFail = function () {
            return view('errors.custom', ['code' => 404, 'message' => 'The site you are trying to access does not exist. Check your link and try again.', 'headline' => 'Site not found']);
        };

        InitializeTenancyBySubdomain::$onFail = function () {
            return view('errors.custom', ['code' => 404, 'message' => 'The subdomain you are trying to access does not exist. Check your link and try again.', 'headline' => 'Subdomain not found']);
        };

        $this->makeTenancyMiddlewareHighestPriority();

        Livewire::setUpdateRoute(function ($handle) {
            return Route::post('/livewire/update', $handle)
                ->middleware([
                        'universal',
                        InitializeTenancyByDomainOrSubdomain::class,
                        'web'
                    ]
                );
        });

        FilePreviewController::$middleware = ['web', 'universal', InitializeTenancyByDomainOrSubdomain::class];
    }
```

Any suggestions to debug this? I really want to get the 404's working for subdomains etc.

**Comment by JeremyRed** [comment]
Commenting here in case anyone else lands on this page, we hit the same issue using Livewire alongside `UniversalRoutes`. You can work around it by creating your own version of the `UniversalRoutes` feature class and slightly modifying the bootstrapped behavior to preserve your own `onFail` logic, rather than having it overridden.

This is based on a closed PR available here: [#1471](https://github.com/archtechx/tenancy/pull/1471)

```php
<?php

namespace App\Support;

use Closure;
use Stancl\Tenancy\Features\UniversalRoutes as BaseUniversalRoutes;
use Stancl\Tenancy\Tenancy;

class UniversalRoutes extends BaseUniversalRoutes
{
    public function bootstrap(Tenancy $tenancy): void
    {
        foreach (static::$identificationMiddlewares as $middleware) {
            // preserve any onFail callback set before universal routes bootstraps
            $customOnFail = $middleware::$onFail;

            $middleware::$onFail = function ($exception, $request, $next) use ($customOnFail) {
                if (static::routeHasMiddleware($request->route(), static::$middlewareGroup)) {
                    return $next($request);
                }

                if ($customOnFail instanceof Closure) {
                    return $customOnFail($exception, $request, $next);
                }

                throw $exception;
            };
        }
    }
}

```

The ordering of the `onFail` override is important which is why the old solution wasn't working. We are using this with Livewire 4 and the latest version of this package.
